### ISSUE TITLE
Singleton Mediator with Scoped context per Handler

### ISSUE THREAD

**Issue by thiagolunardi** [jbogard/MediatR#309]
# User Story

Related to our [previous conversation](https://github.com/jbogard/MediatR.Extensions.Microsoft.DependencyInjection/issues/42), I've been creating my custom implementation of `IMediator` to achieve the following goals:

* ME as a `MediatR` consumer;
* WHEN I send a command request;
* THEN I want to wait for the command request result
* AND I want to return the result as HTTP response, regardless if one or more others command request or event notification was sent or raised;

## TL;DR;

Considering the following scenario:

Action | Reaction
-|-
 send RegisterUserCommand | raise UserRegisteredEvent 
 raise UserRegisteredEvent | send SendConfirmationEmailCommand<br/>send SendAdminNotificationEmailCommand 
 send SendConfirmationEmailCommand | raise ConfirmationEmailSentEvent 
 send SendAdminNotificationEmailCommand | raise AdminNotificationEmailSentEvent 
 raise ConfirmationEmailSentEvent | *do nothing* 
 raise AdminNotificationEmailSentEvent | *do nothing* 

Considering the [current design of MediatR](https://github.com/jbogard/MediatR/wiki#requestresponse), in the following sample:

    public async IAction RegisterUser(UserModel model)
    {
        await _mediator.Send(new RegisterUserCommand(model.UserName, model.Email));
    }

The HTTP Response will only happen when the entire pipeline is finished, and no more command request or event notification is sent or triggered. Something represented by this:

    await RegisterUserCommand
        => await UserRegisteredEvent
            => await SendConfirmationEmailCommand
                => await ConfirmationEmailSentEvent
                => await AdminNotificationEmailSentEvent
    => return HTTP Response

**Important**: MediatR is an amazing project, it's bringing a lot of value to the projects I'm implementing it. It has my full support, admiration, and appreciation.

The need I have here is: I don't want to wait for anything else after the first command request `RegisterUSerCommand` is completed. Anything after that is just reaction, as the main action is already completed, the user is already registered. If something after the persisting of data happens, like any problem sending any of those emails, will not invalidate the user registration action.

By that, I decided to tweak the current implementation of `IMediator`, to able me to send a command request, wait for its completions, return the HTTP Response to the client and leave the reaction chain still running by itself.

At first tries, I got several `ObjectDisposedException` throws, because as the current design of `Mediator` lifetime is per request, after the response, the IoC start to dispose all objects, leading my chain to break because of this. So kept going to tweaking my custom implementation, as a way to achieve two things:

* Make `Mediator` instance a singleton per web host;
* Make every command and event sent or raised have it own IoC scoped context;

The `Mediator` instance must be a singleton to keep and manage all chain tasks started, ensuring that they will be awaited until its completion. And every sent message is creating a single scope of IoC container and resolving its handlers and injected dependencies, just being disposed of after each handler is finished.

Now I can accomplish my need like this:

    await RegisterUserCommand
    => return HTTP Response

And after, my singleton `Mediator` custom implementation handle the rest of the reaction chain, asynchronously, without holding the HTTP Response until its finished.

The main change made was at `Send()` and `Publish()` method, like:

    var taskRunner = Task.Run(async () =>
    {
        using (var scope = _serviceFactory.NewScope())
        {
            var handler = (RequestHandlerWrapper<TResponse>)Activator
                .CreateInstance(typeof(RequestHandlerWrapperImpl<,>).MakeGenericType(requestType, typeof(TResponse)));
            return await handler.Handle(request, cancellationToken, _serviceFactory);
        }
    });

By that, I'd like to do one of two things:
- If it's your interest, to bring this feature/behavior to your `Mediator`, an option way of use, ofc;
- If not your interest, to publish my changes as another component, making all references to your work, naturally;

I just want to share what I achieve, but first I'd like to hear from you if it can be embedded in `MediatR`, or if you agree to be an extension of your project.

Cheers!

**Comment by jbogard** [comment]
Have you thought about creating something above `IMediator` instead? Because I'm not sure your assumption is actually the case, that `Mediator` was designed for any one particular lifecycle. It gets passed a factory method, and you can do whatever you want, no?

**Comment by joaomatossilva** [comment]
To be honest, the part of not wanting to wait for other events to complete is a "Fire and Forget" approach. That approach is very un-recomended on a asp.net application (either core or framework) because the server can stop and the requests are not made. Than can also happen with threads after the request has been finished.

If you need that sort of "fire and forget", you'll be best served with a Message Queuing system. You'll just wait for the message to be en-queued and leave. That should be a small time.

**Comment by thiagolunardi** [comment]
@joaomatossilva you're correct, it's a behavior like "fire and forget" approach the one I'm looking for, but not only this, not as simple as that.

I want to await the completion of the first command request, in my example, to register a user. I want to wait for the user to be registered and as soon as it's completed, raise others events and then leave.

And I want to make this work inside the application pipeline stack.

**Comment by thiagolunardi** [comment]
@jbogard The situation I have right now is that I'm bringing `Mediator` to a already existing system, so all services registrations were made as Scoped. Means I need to create a new scope per each sent message.

**Comment by jbogard** [comment]
Maybe don’t use MediatR?

You might be better off building your own thing, it’s only a couple hundred
lines of code.

**Comment by jbogard** [comment]
No problem!
